### PR TITLE
Multi paste refactor & fix

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
*Pick one, remove the rest*
- Refactoring to reduce code duplication
- It fixes a bug: incorrect memory access when using variable batch size

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * Make the operator CRTP, move common code to the sperclass
     * Separate shape and data handling
     * Remove output-dependent code from Setup
     * Split SetupImpl into common base and backend-dependent SetupTyped
     * Move SetupImpl to superclass
 - Affected modules and functionalities:
     * Multipaste
 - Key points relevant for the review:
     * N/A
 - Validation and testing:
     * Unlocked variable batch size test
     * Make main test deterministic
 - Documentation (including examples):
     * N/A

**JIRA TASK**: DALI-2105

### CODE REVIEW

File: dali/kernels/imgproc/paste/paste_gpu.h [hunk 1]
```diff
@@ -36,6 +37,7 @@ namespace paste {
 template<class InputType, int ndims>
 struct PatchDesc {
   const InputType *in;
+  int in_sample_idx;
```
Comment (mzient) [inline dali/kernels/imgproc/paste/paste_gpu.h:40]: Well, this _probably_ could be skipped or there could be a union in/sample_idx, but that would be uglyish.

**Comment by mzient** [comment]
!build

File: dali/operators/image/paste/multipaste.h [hunk 2]
```diff
@@ -172,6 +179,52 @@ class MultiPasteOp : public Operator<Backend> {
     }
   }
 
+  using Operator<Backend>::RunImpl;
+
+  void RunImpl(workspace_t<Backend> &ws) override {
+    const auto input_type_id = ws.template InputRef<Backend>(0).type().id();
+    TYPE_SWITCH(input_type_id, type2id, InputType, (uint8_t, int16_t, int32_t, float), (
```
Comment (JanuszL) [inline dali/operators/image/paste/multipaste.h:186]: Can we make the supported types a define and list them only once in the file?

**Comment by mzient** [comment]
!build

File: dali/operators/image/paste/multipaste.h (hunk 2 above)
Comment (mzient) [reply dali/operators/image/paste/multipaste.h]: Done (that and more).

**Review by JanuszL — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Pull request merged**